Correct boost in script_score query and error on negative scores

mayya-sharipova opened the pull request:
Before boost in script_score query was wrongly applied only to the subquery.
This commit makes sure that the boost is applied to the whole score
that comes out of script.

Also provide error 400x error message on negative scores in script_score.

Closes #48465

mayya-sharipova committed:
Correct boost calculation in script_score query

Before boost in script_score query was wrongly applied only to the subquery.
This commit makes sure that the boost is applied to the whole score
that comes out of script.

Closes #48465

elasticmachine commented:
Pinging @elastic/es-search (:Search/Search)

matriv commented on docs/reference/query-dsl/script-score-query.asciidoc:
```diff
@@ -48,9 +48,12 @@ scores be positive or `0`.
 
 `min_score`::
-(Optional, float) Documents with a <<relevance-scores,relevance score>> lower
+(Optional, float) Documents with a score lower
```
why did you remove the reference to the score docs?

mayya-sharipova replied:
Before we were referencing relevance scores, while I think the goal of script_score query is to calculate custom scores through some scripts, not a traditional textual relevance score.

matriv commented on server/src/main/java/org/elasticsearch/common/lucene/search/function/ScriptScoreQuery.java:
```diff
@@ -143,7 +148,12 @@ public Explanation explain(LeafReaderContext context, int doc) throws IOExceptio
                     }
                 }
-                
+                if (boost != 1.0f) {
```
This seems to be executed also when `explanation == null`? Or I'm missing something?
Maybe it worths adding a test where the boost is ` != 1` and check that there is no explanation returned regarding the boost?

Also, why not return an explanation if the boost is `1`, since it's asked by the user?

mayya-sharipova replied:
The way we handle `explanation` object is different from other queries, which may be confusing. In `script_score` query a user can provide his/her custom explanation. 

> This seems to be executed also when explanation == null?

Not exactly. We start with `explanation` that is user provided explanation for the script, and if it null, it is substituted by our standard explanation (line 143), so on in line 151 explanation is never null. 

> Maybe it worths checking for a test that when the boost is != 1 and explanation is false there is no explanation returned regarding the boost?

When is explanation is not asked (`explain = false` in a search request),  we will not even go to this method. This method is executed only when a user requests an explanation. 

> Also, why not return an explanation if the boost is 1, since it's asked by the user?

I was thinking since `boost` is an optional parameter,  when it is not provided by a user, there is no need to provide an explanation about it.

matriv commented on modules/lang-painless/src/test/resources/rest-api-spec/test/painless/110_script_score_boost.yml:
```diff
@@ -0,0 +1,86 @@
+              query: {match_all: {boost: 10}}
+              script:
+                source: "doc['i'].value * _score"
+              boost: 10
```
Maybe use a different score here, e.g.: `5` to make it more clear.

matriv reviewed (commented):
Looks good, I'm completely inexperienced in the area but I left some comments, especially the one regarding explanation.

mayya-sharipova committed:
Address feedback

mayya-sharipova commented:
@matriv Thanks for the review. I have tried to address your comments, please continue to review when you have time.

matriv reviewed (approved):
LGTM. Thanks a lot for responding to my questions and providing some more context!

jpountz commented on server/src/main/java/org/elasticsearch/common/lucene/search/function/ScriptScoreQuery.java:
```diff
@@ -203,30 +213,33 @@ public int hashCode() {
+            float score = (float) scoreScript.execute(explanation) * boost;
             if (score == Float.NEGATIVE_INFINITY || Float.isNaN(score)) {
                 throw new ElasticsearchException(
                     "script_score query returned an invalid score [" + score + "] for doc [" + docId + "].");
```
I think it would be better if this error message returned the value produced by the script without the boost.

jpountz commented on server/src/main/java/org/elasticsearch/common/lucene/search/function/ScriptScoreQuery.java:
```diff
@@ -143,7 +148,12 @@ public Explanation explain(LeafReaderContext context, int doc) throws IOExceptio
+                    subs.addAll(Arrays.asList(explanation.getDetails()));
+                    subs.add(Explanation.match(boost, "boost"));
+                    explanation = Explanation.match(explanation.getValue(), explanation.getDescription(), subs);
+                }
```
I'd suggest wrapping the explanation instead of modifying it in-place: create the scorer with boost=1f a couple lines above, and then here:

```java
if (boost != 1f) {
    explanation = Explanation.match(boost * explanation.getValue().floatValue(), "Boosted score, product of:",
        Explanation.match(boost, "boost"),
        explanation);
}
```

jpountz reviewed (approved):
I left a minor suggestion, otherwise LGTM.

mayya-sharipova committed:
Address Feedback

mayya-sharipova commented:
@jpountz Thanks for the feedback, I have addressed it.
In the [last commit](https://github.com/elastic/elasticsearch/pull/52478/commits/96db563f4c5eb0e3c9e694976db0c2b65ed2dd9b#diff-343986e8723b9c666edbb9c2d45dd2f0R233) , I have also corrected the exception to make sure a 400 error is returned when a score is negative.

mayya-sharipova committed:
Remove unused imports

mayya-sharipova committed:
Merge remote-tracking branch 'upstream/master' into script_score_boost

jtibshirani commented on server/src/main/java/org/elasticsearch/common/lucene/search/function/ScriptScoreQuery.java:
```diff
@@ -221,12 +225,13 @@ public float score() throws IOException {
-            if (score == Float.NEGATIVE_INFINITY || Float.isNaN(score)) {
-                throw new ElasticsearchException(
-                    "script_score query returned an invalid score [" + score + "] for doc [" + docId + "].");
+            if (score < 0f || Float.isNaN(score)) {
```
It looks like this PR also fixed a bug in `script_score` queries where we allowed negative scores. I think we should add a note to the breaking changes docs and also update the PR description to make it clear we included this change.

mayya-sharipova replied:
@jtibshirani Thanks.  Even before without this change a user would get an error if their `script_score` query produced a negative score. They would just get it from a different place, I think from the Lucene [here](https://github.com/apache/lucene-solr/blob/master/lucene/core/src/java/org/apache/lucene/search/TopScoreDocCollector.java#L79)

So the only thing changed from a user perspective is an error message and error status code (before was 500, not 400x).  Do you think it warrants a breaking change notice?
+1 for include this in the PR description

jtibshirani replied:
I tried this out using an Elasticsearch 7.6 build, and didn't receive an error:

```
PUT my_index/_doc/1?refresh
{
  "field": "value"
}

GET my_index/_search
{
  "query": {
    "script_score": {
      "query": {
        "match_all": {}
      },
      "script": {
        "source": "-1000"
      }
    }
  }
}
```

The line you linked to is an assert, so perhaps these Lucene checks didn't always catch the issue in non-test environments.

mayya-sharipova replied:
@jtibshirani Thanks for uncovering this.  I understood what happened:
- Before 7.5,  `script_score` query was using [ScriptScoreFunction](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/common/lucene/search/function/ScriptScoreFunction.java#L83) that was returning 400 error with a negative score.  
- From 7.5, we have changed it to not use `ScriptScoreFunction` but forgot to add a condition for a negative score.  But  [TopScoreDocCollector assertion](https://github.com/apache/lucene-solr/blob/master/lucene/core/src/java/org/apache/lucene/search/TopScoreDocCollector.java#L79) is tripped, causing fatal error in the dev mode. But I guess we silence these assertions in a production mode  as we don't see any visible errors or error log messages.

So, negative scores were wrongly allowed only in 7.5-7.6 versions, so to me it doesn't look like  a breaking change. WDYT?

jtibshirani replied:
Adding an explanation to the release notes makes sense to me. I agree it shouldn't be presented as a typical 'breaking change', it is more like a regression that we fixed. Perhaps we could add a unit test along with the release notes update, to prevent a future regression?
